drivers/fpga/fpga_ice40_bitbang.c

```diff
@@ -0,0 +1,289 @@
+				 config_bitbang->clear, cs, clk, pico, (uint8_t *)image_ptr,
+				 img_size);
+
+	LOG_DBG("Send %u clocks", config->trailing_clocks);
+	fpga_ice40_send_clocks(config_bitbang->mhz_delay_count, config_bitbang->set,
+			       config_bitbang->clear, clk, config->trailing_clocks);
+
+	LOG_DBG("checking CDONE");
+	ret = gpio_pin_get_dt(&config->cdone);
+	if (ret < 0) {
+		LOG_ERR("failed to read CDONE: %d", ret);
+		goto unlock;
+	} else if (ret != 1) {
+		ret = -EIO;
+		LOG_ERR("CDONE did not go high");
+		goto unlock;
+	}
+
+	ret = 0;
+	data->loaded = true;
+	fpga_ice40_crc_to_str(crc, data->info);
+	LOG_INF("Loaded image with CRC32 0x%08x", crc);
+
+unlock:
+	(void)gpio_pin_configure_dt(&config->creset, GPIO_OUTPUT_HIGH);
+	(void)gpio_pin_configure_dt(&config->bus.config.cs.gpio, GPIO_OUTPUT_HIGH);
+	(void)gpio_pin_configure_dt(&config_bitbang->clk, GPIO_DISCONNECTED);
+	(void)gpio_pin_configure_dt(&config_bitbang->pico, GPIO_DISCONNECTED);
+#ifdef CONFIG_PINCTRL
+	(void)pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+#endif /* CONFIG_PINCTRL */
+
+	k_spin_unlock(&data->lock, key);
+
+	return ret;
+}
+
+static const struct fpga_driver_api fpga_ice40_api = {
+	.get_status = fpga_ice40_get_status,
+	.reset = fpga_ice40_reset,
+	.load = fpga_ice40_load,
+	.on = fpga_ice40_on,
+	.off = fpga_ice40_off,
+	.get_info = fpga_ice40_get_info,
+};
+
+#ifdef CONFIG_PINCTRL
+#define FPGA_ICE40_PINCTRL_DEFINE(inst) PINCTRL_DT_DEFINE(DT_INST_PARENT(inst))
+#define FPGA_ICE40_PINCTRL_GET(inst)    .pincfg = PINCTRL_DT_DEV_CONFIG_GET(DT_INST_PARENT(inst)),
+#else
+#define FPGA_ICE40_PINCTRL_DEFINE(inst)
+#define FPGA_ICE40_PINCTRL_GET(inst)
+#endif /* CONFIG_PINCTRL */
+
+#define FPGA_ICE40_DEFINE(inst)                                                                    \
+	FPGA_ICE40_ASSERT_RANGE_SPI_MAX_FREQUENCY(inst);                                           \
+	FPGA_ICE40_ASSERT_RANGE_CONFIG_DELAY(inst);                                                \
+	FPGA_ICE40_ASSERT_RANGE_CRESET_DELAY(inst);                                                \
+	FPGA_ICE40_ASSERT_RANGE_LEADING_CLOCKS(inst);                                              \
+	FPGA_ICE40_ASSERT_RANGE_TRAILING_CLOCKS(inst);                                             \
+	BUILD_ASSERT(DT_INST_PROP(inst, mhz_delay_count) >= 0);                                    \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, creset_gpios));                                   \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, cdone_gpios));                                    \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, clk_gpios));                                      \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, pico_gpios));                                     \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, gpios_set_reg));                                  \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, gpios_clear_reg));                                \
```

Comment:
I do not remember seeing a way to specify min/max in devicetree, but `required: True` on the bindings guards against them not being undefined.
No harm in checking twice though.

Reply:
Those are actually not necessary anymore with a required, I will remove them.

Reply:
done